docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+# F# LSP
+
+F# LSP support design proposal. To be expanded as we learn more / settle on things.
+
+## Table of contents
+
+- [F# LSP](#f-lsp)
+  - [Table of contents](#table-of-contents)
+  - [Goals / Requirements](#goals--requirements)
+  - [Overview](#overview)
+  - [Components](#components)
+    - [FsLSP](#fslsp)
+      - [FsLSPServer](#fslspserver)
+      - [Workspace State Management](#workspace-state-management)
+        - [Things to keep track of:](#things-to-keep-track-of)
+      - [Project model](#project-model)
+      - [Example sequence diagram](#example-sequence-diagram)
+      - [Concurrency and cancellation](#concurrency-and-cancellation)
+      - [Request tracking](#request-tracking)
+    - [NewVsix](#newvsix)
+    - [FCS](#fcs)
+  - [LSP library](#lsp-library)
+    - [Candidates](#candidates)
+  - [LSP client (?)](#lsp-client-)
+  - [Diagnostic mode](#diagnostic-mode)
+  - [LSIF](#lsif)
+  - [Performance](#performance)
+  - [Analyzers \& Refactoring support](#analyzers--refactoring-support)
+  - [Mixed F# / C# solutions](#mixed-f--c-solutions)
+  - [Action plan](#action-plan)
+    - [Phase 0](#phase-0)
+    - [Phase 1](#phase-1)
+    - [Phase 2 - 5](#phase-2---5)
+  - [LSP Endpoints](#lsp-endpoints)
+
+
+## Goals / Requirements
+
+- A single LSP server implementation that can be used by both VS and VS Code (or any other consumer)
```

Comment:
I don't really like this wording. I would prefer something like:
One implementation for LSP library, which can be used by all editors, including Visual Studio, which will cover the following:

LSP protocol and extensions handling
Workspace state handling (not the projects handling though, this is external)
Code fixes and refactorings
Delegating LSP if needed (to/from Roslyn)
Analyzers SDK plug-in (?)

Whereas the "server" (hosting it) part is orthogonal, and can either be handled by community projects (Ionide/FSAC) or by our simple LSP host (which we will be using in VS).

Reply:
I updated it. I think we need to say that this is for LSP _server_, because _client_ part won't really be included (or only for VS). But you can use it as a library.